include/hipSYCL/sycl/detail/device_array.hpp

```diff
@@ -124,35 +116,26 @@ struct device_array
   }
 
   HIPSYCL_UNIVERSAL_TARGET
-  bool operator== (const device_array& other) const noexcept
+  bool operator==(const device_array& other) const noexcept
   {
-    for(size_t i = 0; i < N; ++i)
-      if(_data[i] != other._data[i])
-        return false;
-    return true;
+    return _data == other._data;
   }
 
   HIPSYCL_UNIVERSAL_TARGET
-  bool operator!= (const device_array& other) const noexcept
+  bool operator!=(const device_array& other) const noexcept
   {
     return !(*this == other);
   }
 
-  T _data [N];
+  std::array<T, N> _data;
```

Comment:
hmm... The point of `device_array` was to have an `std::array`-like class that can be used in kernels without actually using `std::array`. The problem is that there's no guarantee that `std::array` (or other classes from the STL) avoid C++ features that are not allowed in device code (e.g. because they require a C++ runtime). For example, depending on how it's implemented, even `assert()` might not work.. So, in general I think we should avoid using STL classes in device code like the plague because we cannot really guarantee that they will work for all combinations of supported devices and STL implementations..

Reply:
I changed the implementation to still use a C-style array while resolving the warnings.